### ISSUE TITLE
Issues with rewriting header with no size change

### ISSUE THREAD

**Issue by nasamaher** [nom-tam-fits/nom-tam-fits#43]
Hi,

I need to be writing data to a binary table while also updating primary header keys.  I have nom tam code that does this and has worked for a year or so until today, when I tried using a HIERARCH key for the first time.  Now when I rewrite the header there is corruption.

I have a jUnit test that shows the situation.  Note, I may be approaching the problem completely wrong so please don't hesitate to correct me!  Ugh, attachments aren't supported here so I am pasting the code.  Oh, the "nom_v1_13" package name is to avoid collisions with a library that uses and older nom tam.

Thanks for any tips
Steve

package nomtamtest;

import java.io.IOException;

import nom_v1_13.tam.fits.BasicHDU;
import nom_v1_13.tam.fits.Fits;
import nom_v1_13.tam.fits.FitsException;
import nom_v1_13.tam.fits.FitsFactory;
import nom_v1_13.tam.fits.Header;
import nom_v1_13.tam.util.BufferedFile;

import org.junit.Test;

public class FItsTest
{

    @Test
    public void test() throws FitsException, IOException
    {
        FitsFactory.setUseAsciiTables(false);
        FitsFactory.setLongStringsEnabled(true);
        FitsFactory.setUseHierarch(true);

        /*
         *  Set up initial header
         */
        Fits f = new Fits();
        BasicHDU primaryHdu = FitsFactory.hduFactory(new float[0]);
        primaryHdu.getHeader().addValue("FILTER", "1,2", "no comment");
        
        primaryHdu.getHeader().addValue(
                "HIERARCH.test","a really long line lkjlkljwef lkjwe flkj weflkjw eflkjw eflkjw eflkjw eflkjwe flkwjefl kjweflk jweflkjw eflkj wef",
                "no comment");
//        primaryHdu.getHeader().addValue(
//                "test","a really long line lkjlkljwef lkjwe flkj weflkjw eflkjw eflkjw eflkjw eflkjwe flkwjefl kjweflk jweflkjw eflkj wef",
//                "no comment");
//        
        f.addHDU(primaryHdu);
        String filename = "img.fits";
        BufferedFile bf = new BufferedFile(filename, "rw");
        f.write(bf);

        /*
         * I will also open a data table and write to it when data arrives...
         */
        
        /*
         * Sometimes, given external input, I will rewrite some header fields ....
         */
        
        /*
         * Using original HDU fails with FitsException: Invalid attempt to rewrite
         */
        try
        {
            primaryHdu.addValue("FILTER", "3,4", "no comment");
            primaryHdu.rewrite();
        }
        catch (FitsException e)
        {
            e.printStackTrace();
        }

        /*
         * This method has been working great .. except now I want to 
         * use HIERARCH keyword and when I do the addValue() corrupts
         * one of the HIEARCH keys.  Can be demonstrated by changing keyword to "HIERARCH.test" above.  The code below will then "remove" the "test" keyword (but leave the value)
         */
        Header headerRewriter = new Fits(filename).getHDU(0).getHeader();
        headerRewriter.addValue("FILTER", "4,5", "no comment");
        headerRewriter.rewrite();
        
        bf.close();
    }
}

**Comment by nasamaher** [comment]
Ok, a simpler case and better layout.  

Steve



```java
    @Test
    public void test() throws FitsException, IOException
    {
        FitsFactory.setUseHierarch(true);

        Fits f = new Fits();
        BasicHDU primaryHdu = FitsFactory.hduFactory(new float[0]);
        primaryHdu.getHeader().addValue("FILTER", "1,2", "no comment");
        
        primaryHdu.getHeader().addValue(
                "HIERARCH.test","a really",
                "no comment");
//        primaryHdu.getHeader().addValue(
//                "test","some stuff here",
//                "no comment");

        f.addHDU(primaryHdu);
        String filename = "img.fits";
        BufferedFile bf = new BufferedFile(filename, "rw");
        f.write(bf);
        bf.close();
        f.close();


        /*
         * After the following, HIERARCH.test will be corrupted - "test" will be missing.  Works
         * ok with just "test" as keyword.
         */
        Header headerRewriter = new Fits(filename).getHDU(0).getHeader();
        headerRewriter.addValue("FILTER", "4,5", "no comment");
        headerRewriter.rewrite();

    }
```

**Comment by nasamaher** [comment]
Actually the HIERARCH just can't be read correctly.  Looks like FitsHeaderCardParser.parseCardKey() doesn't handle HIERARCH's.  

We're ready to go to read "HIERARCH.test.this"
![nom1](https://cloud.githubusercontent.com/assets/1706941/9183978/847e6b0e-3f80-11e5-808b-6c1f7b46602b.png)

But this.key has test.this stripped.  (This is what I see in the file using FV as well)
![nom2](https://cloud.githubusercontent.com/assets/1706941/9183985/8e174b18-3f80-11e5-955f-d81ca57714d1.png)

**Comment by ritchieGitHub** [comment]
that sounds like a serous bug, I will take a look this Friday.

**Comment by nasamaher** [comment]
It looks like FitsHeaderCardParser.KEYWORD_PATTERN needs to include lower case - also I think you want to escape the period in the second group:

([A-Z|**a-z**|0-9|_|-]+)([ |\\|\\.]*=?)

**Comment by ritchieGitHub** [comment]
Ah, I understand  that means you use lower case keywords. But as far as I know that's not allowed in the standard. But if the previous version allowed it I suppose we keep that support. Ok if you can test that you fix works, I will trigger a fix release this weekend. @TomMcGlynn do you concurre that this would be worth a 1.13.1 release?

**Comment by nasamaher** [comment]
I'm not sure if the previous version supported it.  And I can switch to
upper-case.  Although I imagine that period needs to be escaped.

By the way, I'm looking at another issue now so you may want to wait on a
release.  When using LongStringsEnabled the continuation mechanism seems to
incorrectly deal with HIEARCH and longer keywords.  It truncates the (long)
value before the closing single quote.  I will send more later.

**Comment by ritchieGitHub** [comment]
ok, I escaped the dot in the regexp but I find it strange that the tests all worked. So I assume the regexep worked as expected even if the expression was formulated wrong. For support of lowercase characters I want to wait for the opinion of @TomMcGlynn after that I will release 1.13.1 

Please test if the SNAPSHOT works as expected.

**Comment by ritchieGitHub** [comment]
reported fixed
